acvm/src/lib.rs

```diff
@@ -54,10 +57,41 @@ pub enum OpcodeResolutionError {
 }
 
 pub trait Backend:
-    SmartContract + ProofSystemCompiler + PartialWitnessGenerator + Default + Debug
+    SmartContract
+    + ProofSystemCompiler
+    + PartialWitnessGenerator
+    + CommonReferenceString
+    + Default
+    + Debug
 {
 }
 
+#[async_trait]
+pub trait CommonReferenceString {
+    /// The Error type returned by failed function calls in the CommonReferenceString trait.
+    type Error: std::error::Error; // fully-qualified named because thiserror is `use`d at the top of the crate
+
+    /// Provides the common reference string that is needed by other traits
+    async fn generate_common_reference_string(
+        &self,
+        circuit: &Circuit,
+    ) -> Result<Vec<u8>, Self::Error>;
+
+    /// Updates a cached common reference string within the context of a circuit
+    ///
+    /// This function will be called if the common reference string has been cached previously
+    /// and the backend can update it if necessary. This may happen if the common reference string
+    /// contains less than the number of points needed for by circuit, or any other checks the backend
+    /// must do.
+    ///
+    /// If the common reference string doesn't need any updates, implementors can return the value passed.
+    async fn update_common_reference_string(
+        &self,
+        common_reference_string: &[u8],
+        circuit: &Circuit,
+    ) -> Result<Vec<u8>, Self::Error>;
```

Comment:
```suggestion
    /// If the common reference string doesn't need any updates, implementors can return the value passed.
    async fn update_common_reference_string(
        &self,
        common_reference_string: Vec<u8>,
        circuit: &Circuit,
    ) -> Result<Vec<u8>, Self::Error>;
```

With the comment in mind, shouldn't we pass ownership over the CRS to the backend? This way it can just hand ownership right back whereas currently it needs to clone the vector.